### PR TITLE
BEP-221: implement cometBFT light block validation

### PR DESCRIPTION
**Pull request by KeefeL**
### Description

This PR implements the [BEP-221: CometBFT Light Block Validation](https://github.com/bnb-chain/BEPs/blob/master/BEP221.md).

### Rationale

Implement the BEP-221

### Example

TestCometBFTLightBlockValidate

### Changes

Notable changes: 
* Add a precompiled contract for validating the CometBFT light blocks

### CODE REVIEW

File: core/vm/contracts_lightclient.go [hunk 1]
```diff
@@ -327,3 +329,37 @@ func keyVerifier(key string) error {
 	}
 	return nil
 }
+
+// cometBFTLightBlockValidate implemented as a native contract. Used to validate the light
+// blocks for CometBFT and its compatible version.
```
Comment (j75689) [inline core/vm/contracts_lightclient.go:334]: Can we mark which version of CometBFT we are using?

**Comment by guagualvcha** [comment]
Please fix the CI check

File: params/config.go [hunk 2]
```diff
@@ -74,6 +74,28 @@ var (
 		Ethash:              new(EthashConfig),
 	}
 
+	// just for prysm compile pass
+	// RopstenChainConfig contains the chain parameters to run a node on the Ropsten test network.
+	RopstenChainConfig = &ChainConfig{
```
[the remark this answers is not in the document]
Comment (KeefeL) [reply params/config.go:79]: When I compile the geth, it tells me that a code piece in prysm lib has used this var, so just add this to make the prsym compile pass.

**Comment by guagualvcha** [comment]
image

This test case is tricky, you may modify it and copy cometbft dependency into the go.mod

File: core/vm/lightclient/v2/lightclient.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+	height := binary.BigEndian.Uint64(input[pos : pos+heightLength])
+	pos += heightLength
+
+	nextValidatorSetHash := input[pos : pos+validatorSetHashLength]
+	pos += validatorSetHashLength
+
+	validatorSetLength := (inputLen - minimumLength) / singleValidatorBytesLength
+	validatorSetBytes := input[pos:]
+	var validatorSet []*types.Validator
```
[the remark this answers is not in the document]
Comment (KeefeL) [reply core/vm/lightclient/v2/lightclient.go]: Fixed.

File: core/vm/lightclient/v2/lightclient.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+		// NOTE: this should always be the last check because untrustedVals can be
+		// intentionally made very large to DOS the light client. not the case for
+		// VerifyAdjacent, where validator set is known in advance.
+		err = block.ValidatorSet.VerifyCommitLight(cs.ChainID, block.Commit.BlockID, block.Height, block.Commit)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	validatorSetChanged := false
```
Comment (j75689) [inline core/vm/lightclient/v2/lightclient.go:121]: line: 121-124 can be simplified to `validatorSetChanged := !(bytes.Equal(cs.ValidatorSet.Hash(), block.ValidatorsHash))`

Comment (KeefeL) [reply core/vm/lightclient/v2/lightclient.go]: Good idea, fixed.

**Review by guagualvcha — APPROVED** [review]

**Review by j75689 — APPROVED** [review]

**Pull request merged**